Creation of the nitsche taylor-couette example (#352)
lethe-cfd/lethe: doc/source/examples/incompressible-flow/2d-taylor-couette-flow-nitsche/2d-taylor-couette-flow-nitsche.rst

========================================================
2D Taylor-Couette Flow Using Nitsche Immersed Boundaries
========================================================

> This is the XXe Lethe example. It revisits the same taylor-couette flow problem in `Example 2 <https://lethe-cfd.github.io/lethe/examples/incompressible-flow/2d-taylor-couette-flow/2d-taylor-couette-flow.html>`, 

Reviewer: Third Lethe example, for now!

Author: Yes I was waiting for this PR to be ready to merge before udating this number! I will keep the comment open

now using immersed boundaries to represent the inner cylinder. This example demonstrates some of the capabilities of Lethe to simulate the flow around complex geometries 

>  meshing them explicitly with a conformal mesh, but instead by using the Nitsche immersed boundary method available within Lethe.

Reviewer: The skipped line seems to create a box in the doc, is it supposed to be like this?
![image](https://user-images.githubusercontent.com/50414572/150169761-38947c17-6a66-42d3-b0a1-c121ca0e1c20.png)

Author: Thanks!

Features
----------------------------------
- Solvers: ``gls_nitsche_navier_stokes_22`` (with Q1-Q1 and Q2-Q1)
- Steady-state problem
- Displays the use of an adaptative mesh refinement using Nitsche immersed boundaries
- Displays the use of the analytical solution to calculate the mesh convergence 
- Displays the calculation of the torque induced by the fluid on a boundary

Location of the examples
------------------------
``examples/incompressible_flow/2d_nitsche_taylor-couette/uniform_nitsche_taylor-couette.prm``
``examples/incompressible_flow/2d_nitsche_taylor-couette/adaptative_nitsche_taylor-couette.prm``

Description of the case
-----------------------

The Taylor-Couette flow is the name of a fluid flow in the gap between two long concentric cylinders with different rotational velocities. 
We simulate the same case as the regular Taylor-Couette flow where the inner cylinder rotates at a constant angular anti-clockwise velocity :math:`\omega`, 
while the outer cylinder is fixed. 
The following figure shows the geometry of this problem and the corresponding boundary conditions. 

> .. image:: images/taylor-couette.png

Author: It was the original name of the image, I didn't change it :) I can change it to be taylor_couette.png

    :alt: The geometry and boundary conditions
    :align: center
    :name: geometry

Parameter file
--------------

We first establish the meshes used for the simulation. Using Nitsche immersed boundaries, two meshes are to be defined : the fluid mesh and the geometry mesh (i.e. the inner cylinder).

Mesh
~~~~~

The ``mesh`` subsection specifies the computational grid:

.. code-block:: text

  subsection mesh
    set type = dealii
    set grid type = hyper_ball
    set grid arguments = 0, 0 : 1 : true
    set initial refinement = 3
  end

> The ``type`` specifies the mesh format used. We use the ``hyper_shell`` mesh generated from the deal.II `GridGenerator <https://www.dealii.org/current/doxygen/deal.II/namespaceGridGenerator.html>`_ . This GridGenerator generates the mesh of the interstice between two cocentric cylinder. The arguments of this grid type are the position of center of the cylinders (``0, 0``), the inner cylinder radius (`0.25`), the outer cylinder radius (`1`) and the number of subdivision in the azimuthal direction (`4`). All arguments are separated by ``:``. We set ``colorize=true`` and this sets the boundary ID of the inner cylinder to ``0`` and of the outer cylinder to ``1``

Reviewer: " [...] mesh generated from the deal.II `GridGenerator <https://www.dealii.org/current/doxygen/deal.II/namespaceGridGenerator.html>`_."
There an extra space between _ and the end of the phrase.
Sorry, my eye caught it quickly :)

"This GridGenerator generates the mesh of the interstice between two concentric cylinders."
Concentric and plural for cylinder.

" [...] and the number of subdivisions in the azimuthal direction [...]"
Subdivisions plural? Not sure :)

Reviewer: ```suggestion
The ``type`` specifies the mesh format used. We use the ``hyper_shell`` mesh generated from the deal.II `GridGenerator <https://www.dealii.org/current/doxygen/deal.II/namespaceGridGenerator.html>`_. This GridGenerator generates the mesh of the interstice between two concentric cylinders. The arguments of this grid type are the position of center of the cylinders (``0, 0``), the inner cylinder radius (`0.25`), the outer cylinder radius (`1`) and the number of subdivisions in the azimuthal direction (`4`). All arguments are separated by ``:``. We set ``colorize=true`` and this sets the boundary ID of the inner cylinder to ``0`` and of the outer cylinder to ``1``
```

Author: awesome thanks :)

.. image:: images/background_mesh_taylor-couette.png
    :alt: The fluid mesh
    :align: center
    :name: fluid

The ``Nitsche`` subsection specifies the solid geometry embedded in the fluid domain. The Nitsche Immersed Boundary (IB) uses particles located at the 
Gauss quadrature points of the immersed mesh to represent the immersed body. For a thorough explanation of this, we refer the reader to `step-70 <https://www.dealii.org/developer/doxygen/deal.II/step_70>` of deal.II.

.. code-block:: text

  subsection nitsche
  set beta = 10
  set number of solids = 1
  subsection nitsche solid 0
	  subsection mesh
		  set type = dealii
		  set grid type = hyper_ball
		  set grid arguments = 0, 0 : 0.25 : true
		  set initial refinement = 6
	  end
	  subsection solid velocity
		  set Function expression = -y ; x
	  end
    set enable particles motion		= false
  end
  set calculate torques on solid = true
  set verbosity = verbose

> end

Reviewer: The last "end" does not appear int the text code-block in the html generated!
I think you need to add an indentation so that the last "end" is within the code-block.

.. image:: images/taylor_couette_with_nitsche_mesh.png
    :alt: The background mesh and the embedded particles
    :align: center
    :name: nitsche_particles

Boundary conditions
~~~~~~~~~~~~~~~~~~~

The ``boundary conditions`` subsection becomes simple since the inner cylinder boundaries were specified in the previous section.

.. code-block:: text

  subsection boundary conditions
    set number                  = 1
      subsection bc 0
          set type              = noslip
      end
  end

First, the ``number`` of boundary conditions to be applied must be specified. For each boundary condition, the ``id`` of the boundary as well as its ``type`` must be specified. The outer cylinder (``0``) is static and, consequently, a ``noslip`` boundary condition is applied. 

Physical properties
~~~~~~~~~~~~~~~~~~~

The analytical solution for the Taylor-Couette problem is only valid at low Reynolds number. We thus set the kinematic viscosity to 1.

.. code-block:: text

  subsection physical properties
    subsection fluid 0
      set kinematic viscosity            = 1.0
    end
  end

FEM interpolation
~~~~~~~~~~~~~~~~~

> Lethe supports the use of arbitrary interpolation order. The :math:`\mathcal{L}^2` norm of the error is :math:`\mathcal{O}\left(h^{n+1} \right)` where :math:`h` is a measure of the element size and `n=1` is the interpolation order of the velocity. However, since the torque applied on the inner cylinder depends on the deviatoric stress tensor, which depends on the velocity gradient, it's error will be :math:`\mathcal{O}(n)`. Taking this into account, we use second order polynomials in this example to obtain higher accuracy on the torque. We specify the interpolation order for both pressure and velocity using the ``FEM`` subsection:

Reviewer: subsection FEM
    set velocity order            = 1
    set pressure order            = 1
end

> Lethe supports the use of arbitrary interpolation order. The :math:`\mathcal{L}^2` norm of the error is :math:`\mathcal{O}\left(h^{n+1} \right)` where :math:`h` is a measure of the element size and `n=1` is the interpolation order of the velocity. However, since the torque applied on the inner cylinder depends on the deviatoric stress tensor, which depends on the velocity gradient, it's error will be :math:`\mathcal{O}(n)`. Taking this into account, we use second order polynomials in this example to obtain higher accuracy on the torque. We specify the interpolation order for both pressure and velocity using the ``FEM`` subsection:

Author: So since using a GLS solver, you can either use a Q1Q1 elements or Q2Q1 elements. If you use Q1Q1 elements, you will reach a second order convergence for velocity, and if using Q2Q1, you will reach a third order (n+1)

> Lethe supports the use of arbitrary interpolation order. The :math:`\mathcal{L}^2` norm of the error is :math:`\mathcal{O}\left(h^{n+1} \right)` where :math:`h` is a measure of the element size and `n=1` is the interpolation order of the velocity. However, since the torque applied on the inner cylinder depends on the deviatoric stress tensor, which depends on the velocity gradient, it's error will be :math:`\mathcal{O}(n)`. Taking this into account, we use second order polynomials in this example to obtain higher accuracy on the torque. We specify the interpolation order for both pressure and velocity using the ``FEM`` subsection:

Author: I added a precision!

.. code-block:: text

    subsection FEM
        set velocity order            = 1
        set pressure order            = 1
    end

.. warning:: 
    With the ``gls_navier_stokes_2d`` solver, Q2-Q2 elements could also be used. However, we have not found that these lead to better results when the flows are at a low Reynolds number.

Analytical solution

> --------------------

Reviewer: Not the good heading I think!
You are still explaining the Parameter file section of your doc, so it needs tilde symbols.
But that is my opinion :)

Author: Oh! The warning was already there just to specify that you can also use this in the FEM section. Maybe this should be moved to the Features section?

Reviewer: As you wish, maybe it is better in the Features because you don't specify anything in the prm file in this section.

Author: Ok I did your suggestion :)

Like in the first Taylor-Couette example, we add an analytical solution section to the parameter handler file. This analytical solution is more complex to define, 
since the simulation domain encompasses the inside of the inner cylinder as well as the gap between the cylinders. Because of this, we only specify the analytical 
solution for the velocity field and forego pressure. The analytical solution is only defined in the ``.prm`` file and we do not reproduce it here for the sake of brevity.

Forces

> -------

Reviewer: Heading with tilde symbols.

Author: Why du you think it should be tild here? As all the other sections are in --

Reviewer: The Boundary conditions, FEM interpolation, Mesh, etc, are all main subsections of your parameter file and they use tilde symbols. Forces also is a main subsection of the parameter file, so thats why I think it should have the same heading!

The ``forces`` subsection controls the postprocessing of the torque and the forces acting on the boundaries of the domain.

.. code-block:: text

  subsection forces
    set verbosity             = verbose   # Output force and torques in log 
    set calculate torques     = true     # Enable torque calculation
    set torque name           = torque    # Name prefix of torque files
    set calculation frequency = 1         # Frequency of the force calculation
    set output frequency      = 1         # Frequency of file update
  end

By setting ``calculate torques=true``, the calculation of the torque resulting from the fluid dynamics physics on every boundary of the domain is automatically calculated. 

> Setting ``verbose=verbose`` will print out the value of the torque calculated for each mesh. 

Reviewer: verbosity=verbose

Reviewer: ```suggestion
Setting ``verbosity=verbose`` will print out the value of the torque calculated for each mesh. 
```

Simulation control and mesh refinement

> --------------------------------------

Reviewer: Heading with tilde symbols.

As stated above, this problem can either be solved using a uniform mesh refinement or using an adaptative mesh refinement

Uniform mesh refinement

> ^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: This heading does not appear in the doc!

The ``simulation control`` subsection controls the flow of the simulation. Two additional parameters are introduced in this example. 

> By setting ``number mesh adapt=4`` we configure to simulation to carry out to solve the fluid dynamics on the mesh and on four(4) subsequently refined mesh. 

Reviewer: "[...] we configure the simulation to carry out to solve the fluid dynamics [...]"

Reviewer: ```suggestion
By setting ``number mesh adapt=4`` we configure the simulation to carry out to solve the fluid dynamics on the mesh and on four(4) subsequently refined mesh. 
```

This approach is very interesting, because the solution on the coarse mesh also serves as the initial guest for the solution on the finer mesh. 

.. code-block:: text

  subsection simulation control
    set method                  = steady
    set number mesh adapt       = 4
    set output name             = taylor_couette_22
    set output frequency        = 1
    set output path             = ./
  end

We then set the mesh adaptation ``type`` to ``uniform``.

.. code-block:: text

  subsection mesh adaptation
    set type                    = uniform
  end

Adaptative mesh refinement

> ^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: This heading does not appear in the doc!

Since the Nitsche IB method introduces additional error on the surface of the immersed geometry, it is pertinent to investigate the results it can produce with 
adaptive mesh refinement. We now consider the following option:

.. code-block:: text

  subsection simulation control
    set method                  = steady
    set number mesh adapt       = 6
    set output name             = taylor_couette_22
    set output frequency        = 1
    set output path             = ./
  end

The mesh can be dynamically adapted using Kelly error estimates on the velocity, pressure or variables arising from other physics. 

.. code-block:: text

  subsection mesh adaptation
    set type                    = kelly
    set variable                = velocity
    set fraction type           = number
    set max number elements     = 500000
    set max refinement level    = 15
    set min refinement level    = 0
    set frequency               = 1
    set fraction refinement     = 0.3
    set fraction coarsening     = 0.15
  end

Rest of the subsections

> ------------------------

Reviewer: Tilde symbols.

The non-linear and linear solvers subsections do not contain any new information in this example.

Running the simulation
----------------------
Launching the simulation is as simple as specifying the executable name and the parameter file. Assuming that the ``gls_nitsche_navier_stokes_22`` executable is within your path, the simulation can be launched by typing:

.. code-block:: text

>   gls_navier_stokes_22 uniform_nitsche_taylor-couette.prm

Reviewer: gls_nitsche_navier_stokes_22 uniform_nitsche_taylor-couette.prm

Reviewer: ```suggestion
  gls_nitsche_navier_stokes_22 uniform_nitsche_taylor-couette.prm
```

or 

.. code-block:: text

>   gls_navier_stokes_22 adaptative_nitsche_taylor-couette.prm

Reviewer: gls_nitsche_navier_stokes_22 adaptative_nitsche_taylor-couette.prm

Reviewer: ```suggestion
  gls_nitsche_navier_stokes_22 adaptative_nitsche_taylor-couette.prm
```

========================================================
2D Taylor-Couette Flow Using Nitsche Immersed Boundaries
========================================================

> This is the XXe Lethe example. It revisits the same taylor-couette flow problem in `Example 2 <https://lethe-cfd.github.io/lethe/examples/incompressible-flow/2d-taylor-couette-flow/2d-taylor-couette-flow.html>`_, 

Author: Yeah I was trying to use the same first sentence. It`s not necessary I'll remove it

    :align: center
    :name: geometry

Parameter file
--------------

We first establish the meshes used for the simulation. Using Nitsche immersed boundaries, two meshes are to be defined : the fluid mesh and the geometry mesh (i.e. the inner cylinder).

Mesh
~~~~~

The ``mesh`` subsection specifies the computational grid:

.. code-block:: text

  subsection mesh
    set type = dealii
    set grid type = hyper_ball
    set grid arguments = 0, 0 : 1 : true
    set initial refinement = 3
  end

The ``type`` specifies the mesh format used. We use the ``hyper_shell`` mesh generated from the deal.II `GridGenerator <https://www.dealii.org/current/doxygen/deal.II/namespaceGridGenerator.html>`_. This GridGenerator generates the mesh of the interstice between two concentric cylinders. The arguments of this grid type are the position of center of the cylinders (``0, 0``), the inner cylinder radius (`0.25`), the outer cylinder radius (`1`) and the number of subdivisions in the azimuthal direction (`4`). All arguments are separated by ``:``. We set ``colorize=true`` and this sets the boundary ID of the inner cylinder to ``0`` and of the outer cylinder to ``1``

> The last parameter specifies the ``initial refinement`` of the grid. Most deal.II grid generators contain a minimal number of cells. The *hyper_shell* mesh is made of four cells. Indicating an ``initial refinement=3`` implies that the initial mesh is refined 3 times. In 2D, each cell is divided by 4 per refinement. Consequently, the final grid is made of 256 cells.

Author: I had not noticed! I changed it :)

.. image:: images/taylor_couette_with_nitsche_mesh.png
    :alt: The background mesh and the embedded particles
    :align: center
    :name: nitsche_particles

Boundary conditions
~~~~~~~~~~~~~~~~~~~

The ``boundary conditions`` subsection becomes simple since the inner cylinder boundaries were specified in the previous section.

.. code-block:: text

  subsection boundary conditions
    set number                  = 1
      subsection bc 0
          set type              = noslip
      end
  end

First, the ``number`` of boundary conditions to be applied must be specified. For each boundary condition, the ``id`` of the boundary as well as its ``type`` must be specified. The outer cylinder (``0``) is static and, consequently, a ``noslip`` boundary condition is applied. 

Physical properties
~~~~~~~~~~~~~~~~~~~

The analytical solution for the Taylor-Couette problem is only valid at low Reynolds number. We thus set the kinematic viscosity to 1.

.. code-block:: text

  subsection physical properties
    subsection fluid 0
      set kinematic viscosity            = 1.0
    end
  end

FEM interpolation
~~~~~~~~~~~~~~~~~

> Lethe supports the use of arbitrary interpolation order. The :math:`\mathcal{L}^2` norm of the error is :math:`\mathcal{O}\left(h^{n+1} \right)` where :math:`h` is a measure of the element size and `n=1` is the interpolation order of the velocity. However, since the torque applied on the inner cylinder depends on the deviatoric stress tensor, which depends on the velocity gradient, it's error will be :math:`\mathcal{O}(n)`. Taking this into account, we can use second order polynomials in this example to obtain higher accuracy on the torque. We specify the interpolation order for both pressure and velocity using the ``FEM`` subsection:

Author: Good point I missed it.

The torque on the inner cylinder should be -0.83776 and we note that the torque on both cylinder converges close to that value. 
Running the simulation with finer meshes lead to this results.

Adaptative mesh refinement
^^^^^^^^^^^^^^^^^^^^^^^^^^

Using Paraview, the steady-state velocity profile can be visualized for the adaptative mesh refinement case:

.. image:: images/taylor_couette_with_nitsche_velocity_adapt.png
    :alt: velocity distribution
    :align: center

As can be seen, each cell is curved because a Q2 isoparametric mapping was used (by setting ``qmapping=true`` in the FEM subsection). 
To visualize these high-order cells, we need to subdivide the regular cell to store additional information onto them. 
By default, Paraview does not render high-order elements. 
To enable the rendering of high-order elements, the Nonlinear subdivision level slider must be increased above one. For more information on this topic, 
please consult the deal.II wiki page on `rendering high-order elements <https://github.com/dealii/dealii/wiki/Notes-on-visualizing-high-order-output>`.

> 

Reviewer 2: This part should be removed :)
